spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,426 @@
+package org.apache.spark.sql.cassandra
+
+import scala.reflect.ClassTag
+
+import org.apache.spark.Logging
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.cassandra.CassandraSQLRow.CassandraSQLRowReader
+import org.apache.spark.sql.catalyst.expressions.{Literal, AttributeReference, Expression, Attribute}
+import org.apache.spark.sql.sources._
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{DataFrame, Row, SQLContext}
+
+import com.datastax.spark.connector._
+import com.datastax.spark.connector.cql.{CassandraConnector, ColumnDef}
+import com.datastax.spark.connector.rdd.{ReadConf, ValidRDDType}
+import com.datastax.spark.connector.writer.{WriteConf, SqlRowWriter}
+
+/**
+ * Table scan type. It allows to create a Relation for given table, keyspace, cluster and user
+ * defined schema. It also allow to pass [[CassandraConnector]], [[ReadConf]] and [[WriteConf]]
+ * explicitly
+ */
+sealed trait ScanType {
```

Comment:
Filter is a little bit higher level  on top Expression. I don't think that "Both scans hook into exactly the same place in Spark code-base."

First it doesn't do any transfer from Expression to Filter which is more direct access to planner.  Second our combined filter is constructed to directly use Expression internal eval which may have some performance gain as well.

Second, it may have more filtering or different filter path then PrunedFilterScan (need do some investigation)


If there is no gain why have it CatalystScan there?
some internal class uses it https://github.com/apache/spark/blob/02ec058efe24348cdd3691b55942e6f0ef138732/sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala#L84